Expand indirection cell calls in IR instead of doing ad-hoc optimizations in codegen

jakobbotsch opened the pull request:
When the JIT calls a function that requires an indirection cell argument
it is beneficial to load the call target through the already placed arg.
Previously this was done as an ad-hoc optimization during codegen (with
LSRA making sure to allocate an extra register). Instead of this, expand
it as regular IR and remove the ad-hoc optimization.

This change makes the optimization that were previously done work for
CFG as well and generally seems to have (albeit small) positive diffs.

Fix #65076

jakobbotsch committed:
Expand indirection cell calls as IR in lowering

When the JIT calls a function that requires an indirection cell argument
it is beneficial to load the call target through the already placed arg.
Previously this was done as an ad-hoc optimization during codegen (with
LSRA making sure to allocate an extra register). Instead of this, expand
it as regular IR and remove the ad-hoc optimization.

This change makes the optimization that were previously done work for
CFG as well and generally seems to have (albeit small) positive diffs.

Fix #65076

jakobbotsch committed:
Fix build

jakobbotsch committed:
Fix VN of helper calls with indir cells

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
Fixes to build

- Helpers can use indir cell args so we do not only see CT_USER_FUNC
- Fix an incorrect redefinition

AndyAyersMS commented on src/coreclr/jit/gentree.h:
```diff
@@ -4683,47 +4684,15 @@ struct GenTreeCall final : public GenTree
-
-        return NonStandardArgKind::None;
-    }
+    bool HasIndirectionCellArg(fgArgTabEntry** entry) const;
```
Consider adding a default arg value so callers who don't care about the entry don't need to explicitly pass `nullptr`.

```suggestion
    bool HasIndirectionCellArg(fgArgTabEntry** entry = nullptr) const;
```

AndyAyersMS commented on src/coreclr/jit/gentree.h:
```diff
@@ -4683,47 +4684,15 @@ struct GenTreeCall final : public GenTree
         // we need to pass indirection cells as those go into registers that
         // are clobbered by the dispatch helper.
-        bool mayUseDispatcher    = GetIndirectionCellArgKind() == NonStandardArgKind::None;
+        bool mayUseDispatcher    = !HasIndirectionCellArg(nullptr);
```
If you add the default as above:
```suggestion
        bool mayUseDispatcher    = !HasIndirectionCellArg();
```

AndyAyersMS commented on src/coreclr/jit/valuenum.cpp:
```diff
@@ -9759,22 +9759,14 @@ void Compiler::fgValueNumberHelperCallFunc(GenTreeCall* call, VNFunc vnf, ValueN
 
-#if defined(FEATURE_READYTORUN) && defined(TARGET_ARMARCH)
-    if (call->IsR2RRelativeIndir())
+    if (call->HasIndirectionCellArg(nullptr))
```
And here
```suggestion
    if (call->HasIndirectionCellArg())
```

AndyAyersMS reviewed (approved):
LGTM.

Seems like arm is not yet happy though.

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into expand-indirection-cell-calls-as-ir

jakobbotsch committed:
Fix merge

jakobbotsch committed:
Properly kill callee-saved R2R indir cell arg register for calls

The callee-saved register used for the indir cell arg was not being
killed at the call site leaving the arg register marked as busy.

jakobbotsch committed:
Relax CSE of indir cell addresses

After previous fix I believe this should no longer be necessary.

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
Address feedback

jakobbotsch commented:
The ARM issue seemed to be an LSRA bug: for R2R calls we were not killing the register containing the indirection cell address at the call site which led to that register being marked busy for the remainder of the function. ARM is a bit special there since it uses a callee-preserved register. There was a hack in morph to work around this problem (I assume) that I have removed as well.

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into expand-indirection-cell-calls-as-ir

jakobbotsch commented:
On x64 with CFG it looks like it's better to use the constant in both places:
```diff
 ; MyBenchmarks.SimpleIfaceCall.Call()
        push      rax
        xor       eax,eax
        mov       [rsp],rax
        mov       rcx,[rcx+8]
        mov       [rsp],rcx
-       mov       r10,7FF972ED0510
-       mov       rcx,[r10]
+       mov       rcx,[7FF9736E0510]
        call      CORINFO_HELP_VALIDATE_INDIRECT_CALL
        mov       rax,rcx
        mov       rcx,[rsp]
-       mov       r11,r10
+       mov       r11,7FF972EA0510
        cmp       [rcx],ecx
        add       rsp,8
        jmp       rax
-; Total bytes of code 52
+; Total bytes of code 53
 ```
we get
``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.1526 (21H2)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK=7.0.100-alpha.1.21558.2
  [Host]     : .NET 6.0.0 (6.0.21.48005), X64 RyuJIT
  Job-DBANTR : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT
  Job-BFNMJK : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT

EnvironmentVariables=COMPlus_JitForceControlFlowGuard=1  

```
| Method |        Job |                 Toolchain |     Mean |     Error |    StdDev |   Median | Ratio | RatioSD | Code Size |
|------- |----------- |-------------------------- |---------:|----------:|----------:|---------:|------:|--------:|----------:|
|   Call | Job-DBANTR |    \Core_Root\corerun.exe | 1.599 ns | 0.0778 ns | 0.2271 ns | 1.510 ns |  1.21 |    0.23 |      52 B |
|   Call | Job-BFNMJK | \Core_Root_am\corerun.exe | 1.332 ns | 0.0604 ns | 0.1602 ns | 1.288 ns |  1.00 |    0.00 |      53 B |

where the _am version is the diff version. Will disable the expansion with locals for x64 CFG calls with validator.

jakobbotsch committed:
Disable expansion for x64 CFG calls using validator

jakobbotsch commented:
Not totally sure if this change can replace the old ad-hoc optimization. A few observations:
1. In large functions with many locals the new indir cell locals may not be tracked/may cause other locals to not be tracked, which lead to regressions. It will probably help to reuse the local.
2. In tier0/debug codegen the new locals cause extra spilling, which is to be expected. Not sure if this is really an issue (this is the cause of the aspnet regressions, which seem to be in tier0 method contexts when spot checking)
3. On ARM32 LSRA seems to be bad at dealing with the new locals (and new local when reusing it). Still have not been able to pinpoint one particular issue, it seems the new local leads to globally different register allocation which causes regressions.

jakobbotsch committed:
Reuse local for indir cell addr, assign register early for ARM32

jakobbotsch commented:
Closing this, don't think we can teach RA/liveness to deal well with the extra locals required for this, and the regressions seem too large to ignore.
